### ISSUE TITLE
Demo test_teapot_reflectance.py is not working correctly

### ISSUE THREAD

**Issue by swordigo1995** [BachiLi/redner#26]
I find 'test_teapot_reflectance.py' cannot generate a correct optimization result using the latest release code.
But i remember i can get correct optimization result in early release.

So, does anyone know how to make 'test_teapot_reflectance.py' work again?

**Comment by budmonde** [comment]
What was the earliest release where you had this working?

**Comment by BachiLi** [comment]
I rolled back to the first commit and it was not working (it was certainly working when I produced the results for the paper though!) : (
I suspect this is due to the specular highlights that makes either the loss function sharper and more difficult to optimize, or it makes gradient sampling more noisy. When I change the target roughness to 0.2 it optimizes fine. I will investigate more and come back to you.

**Comment by swordigo1995** [comment]
I have tested the code in commit ba37316502bfb3f16173b2414b153c02fb6c0b9c 'add two sided light source' and it works ! @budmonde @BachiLi

**Comment by BachiLi** [comment]
Seems to be caused by variance in the gradient code. I brought back the hierarchical edge sampling method mentioned in the paper and added gradient clipping to optimization. It works on my end now. Let me know if there are any issues.

**Comment by swordigo1995** [comment]
It works. Thanks a lot !
